openwisp_controller/config/models.py

```diff
@@ -107,6 +108,23 @@ def get_temp_config_instance(self, **options):
         c.device = self
         return c
 
+    def save(self, *args, **kwargs):
+        if not self.key:
+            if app_settings.CONSISTENT_REGISTRATION:
+                if app_settings.HARDWARE_ID_ENABLED:
+                    keybase = self.hardware_id
+                else:
+                    keybase = self.mac_address
+                org_settings = OrganizationConfigSettings \
+                    .objects.get(organization=self.organization)
+                m = hashlib.md5()
+                m.update((keybase + "+" + org_settings.shared_secret)
+                         .encode('utf-8'))
+                self.key = m.hexdigest()
+            else:
+                self.key = get_random_key()
+        super().save(*args, **kwargs)
```

Comment:
this is not python 2 compatible, the versions we're working with are still python 2 compatible, so you will have to fix this.